### ISSUE TITLE
2014.7.1: salt/states/saltmod using incorrect return dict for orchestrate

### ISSUE THREAD

**Issue by aurynn** [saltstack/salt#20615]
```    
Function: salt.state
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/local/opt/pyenv/versions/salt/lib/python2.7/site-packages/salt/state.py", line 1529, in call
                  **cdata['kwargs'])
                File "/usr/local/opt/pyenv/versions/salt/lib/python2.7/site-packages/salt/states/saltmod.py", line 250, in state
                  m_ret = mdata['ret']
              KeyError: 'ret'
     Started: 14:05:12.253806
    Duration: 21406.003 ms
     Changes:
```

I'm chasing using Orchestrate-over-SSH, and it looks like `salt/states/saltmod.py` is using the incorrect return dict structure:

https://github.com/saltstack/salt/blob/2014.7/salt/states/saltmod.py#L249

In the above error, changing L249 (and all `'ret:'` entries in saltmod) allows an orchestration-over-ssh to complete normally.

**Comment by rallytime** [comment]
@aurynn Thanks for the report and for finding that fix. We would be happy to accept a pull request with your proposed fixed, if you feel inclined. If you're willing, be sure to make the change against the 2014.7 branch and submit the pull request against that branch as well. If not, we can circle back around and fix this up as well.

**Comment by warthog** [comment]
It looks like in 2015.5.1 this was fixed in saltmod.py for state() but not function().  Here is a gist with some info of what I found when doing salt.state vs salt.function in an orchestrate-over-ssh setup:

https://gist.github.com/warthog/289621004d6cfc065493

In 2015.5.1 saltmod.py has this in state():

        if 'return' in mdata and 'ret' not in mdata:
            mdata['ret'] = mdata.pop('return')

But function() doesn't have that check:

    for minion, mdata in six.iteritems(cmd_ret):
        m_ret = False
        if mdata.get('retcode'):
            ret['result'] = False
            fail.add(minion)
        if mdata.get('failed', False):
            m_func = False
        else:
            m_ret = mdata['ret']
            m_func = (not fail_function and True) or __salt__[fail_function](m_ret)

**Comment by warthog** [comment]
I can also confirm that editing the block in function() to this makes it all work (although there is probably a cleaner way to do this):

for minion, mdata in six.iteritems(cmd_ret):
    m_ret = False
    if mdata.get('retcode'):
        ret['result'] = False
        fail.add(minion)
    if mdata.get('failed', False):
        m_func = False
    else:
        if 'return' in mdata and 'ret' not in mdata:
            mdata['ret'] = mdata.pop('return')
        m_ret = mdata['ret']
        m_func = (not fail_function and True) or __salt__[fail_function](m_ret)

**Comment by rallytime** [comment]
@warthog Would you be able to submit a pull request with that change? We'd happily review it. If you're willing, please submit the change to the 2015.5 branch.

**Comment by warthog** [comment]
@rallytime Unfortunately this doesn't fix the issue 100%  It did fix it for my simple example using cmd.run, but when trying to use salt.function and pass in - arg and -kwarg salt: True breaks again.

Here is a gist showing this (running with a saltmod.py that has the above "fix" in for mdata['ret'] using user.add as an example:

https://gist.github.com/warthog/e72e4c8567234826ccb8

As I started looking at this I thought it was perhaps something wrong with mdata['retcode'] because the .get() was throwing an error, but on closer inspection I believe this issue to be with /usr/lib/python2.7/dist-packages/salt/client/ssh/__init__.py either not parsing the argv/kwargs properly or not getting them passed in correctly.  But in either event things are still broken when trying to leverage - arg and -kwarg in a salt.function state.orch run.

Here is the specific traceback (also in the gist linked above)

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 335, in handle_routine
    stdout, stderr, retcode = single.run()
  File "/usr/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 707, in run
    stdout, stderr, retcode = self.cmd_block()
  File "/usr/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 946, in cmd_block
    log.debug('Performing shimmed, blocking command as follows:\n{0}'.format(' '.join(self.argv)))
TypeError: sequence item 2: expected string, dict found
[ERROR   ] Target 'test02-ds01.test02.testdom.local' did not return any data, probably due to an error.
[ERROR   ] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1563, in call
    **cdata['kwargs'])
  File "/usr/lib/python2.7/dist-packages/salt/states/saltmod.py", line 402, in function
    if mdata.get('retcode'):
AttributeError: 'str' object has no attribute 'get'
```

If I can get some time to dig into this particular piece of it I will try to provide a fix, but to be honest I'm not that familiar with the inner workings of salt at the moment and tracing calls between all the different functions is going to take some time and effort so I can't promise anything right now :(

**Comment by basepi** [comment]
I'm guessing that it's argument passing or similar. This is what's actually in mdata:

```
Target 'basepi02' did not return any data, probably due to an error.
```

basepi02 was my salt-ssh minion in question.

I'm still investigating why it's failing.

**Comment by basepi** [comment]
I'm getting closer. I now have it running successfully, but orchestrate isn't parsing the result successfully:

```
li1293-62.members.linode.com_master:
    ----------
    salt_|-user_add_test_|-user.add_|-function:
        ----------
        __run_num__:
            0
        changes:
            ----------
            out:
                highstate
            ret:
                ----------
                basepi02:
                    False
        comment:
            Running function user.add failed on minions: basepi02 Function user.add ran on basepi02.
        duration:
            1114.133
        name:
            user.add
        result:
            False
        start_time:
            00:54:14.363588
```

**Comment by basepi** [comment]
Oops, nevermind, it's actually working, I had just added the user on a previous try, so it failed the second time (which I think is expected behavior for user.add?)

**Comment by basepi** [comment]
Fixed in #24588.

Keep me posted if that fix doesn't work for you. Should be in 2015.5.3.
